### PR TITLE
Add draft for RFD/0090 - DB MFA Sessions

### PR DESCRIPTION
**Pull request by GavinFrazar**
This RFD describes changes proposed to support per-session-mfa support with `tsh proxy db --tunnel` (which is necessary for use with GUI clients, so effectively this RFD is to support per-session-mfa with GUI clients).

Relevant issue: https://github.com/gravitational/teleport/issues/12538

### CODE REVIEW

File: rfd/0090-db-mfa-sessions.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+and a user may be prompt once per minute for their MFA.
+
+We should remove this 1 minute restriction on database cert TTL to improve UX.
+
+## Details
+
+"Doesn't this just disable per-session-mfa for database access?" (My initial thinking)
+
+- No, not quite. Sessions are still limited by the lifetime of a local proxy process and an MFA prompt is always required to start these proxies when per-session-mfa is enabled.
+- This can "weaken" per-session-mfa security though. If a user finds that they would prefer the old behavior of 1-minute TTL certs, they can still achieve that through configuration using `max_session_ttl`.
```
Comment (zmb3) [inline rfd/0090-db-mfa-sessions.md:35]: Is this true? If "disconnect on expired cert" is set, the user would be disconnected after 1 minute.

File: rfd/0090-db-mfa-sessions.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+---
+ 
+# RFD 90 - Database MFA Sessions
+
+## Required Approvers
+- Engineering: Marek `@smallinsky`, Roman `@r0mant`
+- Product: Sasha `@klizhentas`, Xin `@xinding33`
+
+## What
+When per-session-MFA is enabled, we should not restrict database cert TTL to 1 minute.
```
Comment (zmb3) [inline rfd/0090-db-mfa-sessions.md:13]: This is a little too low-level of an opening sentence. IMHO, the real "what" is:

Prevent database access users from being required to tap an MFA device every minute when per-session-MFA is enabled.

File: rfd/0090-db-mfa-sessions.md (hunk 1 above)
Comment (GavinFrazar) [reply rfd/0090-db-mfa-sessions.md:35]: `disconnect_expired_cert: true` also disconnects per-session-mfa certs after 1 minute, so there's no difference in behavior there. I just tested this with ssh and database access to be sure.

File: rfd/0090-db-mfa-sessions.md (hunk 2 above)
Comment (GavinFrazar) [reply rfd/0090-db-mfa-sessions.md:13]: I agree - moving this low-level description under "Details" section and replacing with your suggestion.

File: rfd/0090-db-mfa-sessions.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+Currently, database certs are issued with 1 minute ttl if per-session-MFA is enabled as described in [RFD 14](https://github.com/gravitational/teleport/blob/master/rfd/0014-session-2FA.md).
+This causes poor UX when the GUI client can establish a new connection at any time,
+and a user may be prompt once per minute for their MFA.
+
+We should remove the 1 minute restriction on database cert TTL to improve UX.
+
+## Details
+When per-session-MFA is enabled, we should not restrict database cert TTL to 1 minute.
+
+Instead, database cert TTL should be restricted to `max_session_ttl`, and the cert
```
Comment (smallinsky) [inline rfd/0090-db-mfa-sessions.md:30]: I think that using current default behavior where cert expiration time is set based on `max_session_ttl` is `reasonable`. Alternatively the DB MFA certs expiration time can be unset allowing to run tsh proxy db forever without second MFA prompt though we will ignore basic teleport cluster restriction in this case.

File: rfd/0090-db-mfa-sessions.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+Currently, `tsh db connect` starts a local proxy if any of the following are true:
+1. TLS routing is enabled.
+2. Proxying connection to a Microsoft SQL server
+3. Proxying connection to a Snowflake database. 
+
+The local proxy it starts only uses a tunnel for Snowflake and Elastisearch.
+
+Instead, we should always use a local proxy and always with a tunnel.
+This way database certs can be held in-memory.
```
[the remark this answers is not in the document]
Comment (GavinFrazar) [reply rfd/0090-db-mfa-sessions.md:75]:
```suggestion
Instead, `tsh db connect` should always use a local proxy and always with a tunnel.
This way database certs can be held in-memory.
```
Sure, how's this

File: rfd/0090-db-mfa-sessions.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+## Required Approvers
+- Engineering: Marek `@smallinsky`, Roman `@r0mant`
+- Product: Sasha `@klizhentas`, Xin `@xinding33`
+
+## What
+When per-session-MFA is enabled, we should not restrict database cert TTL to 1 minute.
+
+Instead, database cert TTL should be restricted to `max_session_ttl`, and the cert
+should be kept in-memory by a local proxy tunnel.
+
```
[the remark this answers is not in the document]
Comment (GavinFrazar) [reply rfd/0090-db-mfa-sessions.md:14]: Added

File: rfd/0090-db-mfa-sessions.md (hunk 3 above)
Comment (ravicious) [reply rfd/0090-db-mfa-sessions.md:30]: Yes, in Connect when the tsh daemon process becomes aware of this, it can notify the Electron app about it and then the app can just bring [its window to the top](https://www.electronjs.org/docs/latest/api/browser-window#winfocus). This is how we decided to solve it for now, not only for per session MFA but also for things like trying to connect to a proxy after your cert has expired.

**Review by xinding33 — APPROVED** [review]
This is a clear improvement to user experience so no objections from me. The biggest open question is definitely security, so as long as the platform security team is ok with the changes, then I'm a +1.

**Review by smallinsky — APPROVED** [review]

**Comment by GavinFrazar** [comment]
@r0mant @klizhentas can you guys take a look? Marek suggested I put you two down as required approvers.

File: rfd/0090-db-mfa-sessions.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+Currently, database certs are issued with 1 minute ttl if per-session-MFA is enabled as described in [RFD 14](https://github.com/gravitational/teleport/blob/master/rfd/0014-session-2FA.md).
+This causes poor UX when the GUI client can establish a new connection at any time,
+and a user may be prompt once per minute for their MFA.
+
+We should remove the 1 minute restriction on database cert TTL to improve UX.
+
+## Details
+When per-session-MFA is enabled, we should not restrict database cert TTL to 1 minute.
+
+Instead, database cert TTL should be restricted to `max_session_ttl`, and the cert
```
Comment (r0mant) [inline rfd/0090-db-mfa-sessions.md:30]: Just to clarify - this will only apply to certificates used by `tsh proxy db`, correct? Certificates produced by `tsh db login` (i.e. those that are dumped to disk) will still be valid for 1 minute? Let's call this out explicitly here.

File: rfd/0090-db-mfa-sessions.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+should be kept in-memory by a local proxy tunnel.
+
+"Doesn't this just disable per-session-mfa for database access?" (My initial thinking)
+
+- No, not quite. Sessions are still limited by the lifetime of a local proxy process and an MFA prompt is always required to start these proxies when per-session-mfa is enabled.
+- This can "weaken" per-session-mfa security though. If a user finds that they would prefer the old behavior of 1-minute TTL certs, they can still achieve that through configuration using `max_session_ttl`.
+
+### API
+
+We should change the auth service to always issue database certs without capping the cert TTL to 1 minute for per-session-MFA.
```
Comment (r0mant) [inline rfd/0090-db-mfa-sessions.md:40]: See my comment above. I think it should only apply to certificates used by the local proxy (which it'll keep in memory).

File: rfd/0090-db-mfa-sessions.md (hunk 6 above)
Comment (r0mant) [reply rfd/0090-db-mfa-sessions.md:30]: Ah, I see below that `tsh db login` will be disabled in per-session MFA mode? Makes sense if so, but let's point it out explicitly here as well.

File: rfd/0090-db-mfa-sessions.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+
+### Local Proxy With Tunnel
```
Comment (r0mant) [inline rfd/0090-db-mfa-sessions.md:42]: Can you add a section somewhere explaining how this can be used with hardware-backed private keys @Joerger has implemented? RFD: https://github.com/gravitational/teleport/pull/14774.

File: rfd/0090-db-mfa-sessions.md [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+Instead, database cert TTL should be restricted to `max_session_ttl`, and the cert
+should be kept in-memory by a local proxy tunnel.
+
+"Doesn't this just disable per-session-mfa for database access?" (My initial thinking)
+
+- No, not quite. Sessions are still limited by the lifetime of a local proxy process and an MFA prompt is always required to start these proxies when per-session-mfa is enabled.
+- This can "weaken" per-session-mfa security though. If a user finds that they would prefer the old behavior of 1-minute TTL certs, they can still achieve that through configuration using `max_session_ttl`.
+
+### API
```
Comment (r0mant) [inline rfd/0090-db-mfa-sessions.md:38]: Can you add a section explaining how it will work with Teleport Connect?

File: rfd/0090-db-mfa-sessions.md (hunk 7 above)
Comment (GavinFrazar) [reply rfd/0090-db-mfa-sessions.md:40]: I can point out here as well that the intention is to disable `tsh db login` and `tsh proxy db` (without --tunnel) when per-session-mfa is required.  So the auth service will "always" issue db certs without capping the TTL to 1 minute, but the client will also always use a proxy tunnel when per-session-mfa is required.

We could have the client send its intent along with the cert request, as in "I intend to keep these certs in memory" or "on disk", but I don't think it's worth doing if we disable non-tunnel clients in per-session-mfa.

So I'll rephrase this to:

```diff
### API

We should change the auth service to always issue database certs without capping the cert TTL to
1 minute for per-session-MFA; clients will only keep the newly issued database certs in-memory
when per-session-mfa is required.
```

File: rfd/0090-db-mfa-sessions.md (hunk 9 above)
Comment (GavinFrazar) [reply rfd/0090-db-mfa-sessions.md:54]: Added

File: rfd/0090-db-mfa-sessions.md (hunk 8 above)
Comment (GavinFrazar) [reply rfd/0090-db-mfa-sessions.md:60]: Added with some detail about integration and security gains (and limitations)

File: rfd/0090-db-mfa-sessions.md (hunk 6 above)
Comment (GavinFrazar) [reply rfd/0090-db-mfa-sessions.md:30]: Added to this section that we will disable these commands for per-session-mfa

**Comment by zmb3** [comment]
See also https://github.com/gravitational/teleport/issues/7568 which is a similar request but for k8s sessions instead of DB sessions.

**Comment by GavinFrazar** [comment]
> See also #7568 which is a similar request but for k8s sessions instead of DB sessions.

While it is a similar request, we don't have an equivalent to `tsh proxy db` for kube access. I think we should have a separate RFD for solving that issue.

File: rfd/0090-db-mfa-sessions.md [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+Database access via GUI client such as DataGrip requires the user to start a local proxy to connect to: `tsh proxy db --tunnel <db>`.
+
+To support per-session-MFA for the local proxy, we are adding a callback function called on each new connection to the local proxy.
+This callback function will check if the database certs need to be reissued to proxy the connection, and prompt for credentials/MFA if needed.
+
+Currently, database certs are issued with 1 minute ttl if per-session-MFA is enabled as described in [RFD 14](https://github.com/gravitational/teleport/blob/master/rfd/0014-session-2FA.md).
+This causes poor UX when the GUI client can establish a new connection at any time,
+and a user may be prompt once per minute for their MFA.
+
+We should remove the 1 minute restriction on database cert TTL to improve UX.
```
[the remark this answers is not in the document]
Comment (GavinFrazar) [reply rfd/0090-db-mfa-sessions.md:25]: `tsh proxy db --tunnel` checks cert expiration now and prompts for credentials/mfa as needed to reissue certs - I added that a few weeks ago.
>Instead of removing the 1 minute TTL on the cert, could we instead perform the per-session MFA reissue when you run tsh proxy db --tunnel <db>?

If the cert TTL is 1 minute, then the local proxy will need to refresh its cert after 1 minute as well. Each incoming connection to the local proxy requires the local proxy to establish a new connection to the remote proxy.

File: rfd/0090-db-mfa-sessions.md [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+Right now, we load certs from the filesystem when using `tsh proxy db --tunnel`.
+So this change will affect per-session-MFA UX flow in that `tsh proxy db --tunnel` will always prompt for MFA when the local proxy starts.
+
+The local proxy can also refresh its certs as needed by invoking a callback function on each new connection which checks the local proxy's db cert expiry.
+
+### Local Proxy Without Tunnel
+
+If per-session-MFA is required, then we should forbid `tsh proxy db <database>` and give the user an error message telling them to add the `--tunnel` flag since per-session-MFA is required.
+
+This prevents per-session-MFA database certs from being saved to the filesystem and directs users towards the better UX of using a local proxy tunnel.
```
[the remark this answers is not in the document]
Comment (GavinFrazar) [reply rfd/0090-db-mfa-sessions.md:74]: I hadn't considered that. We may need to add a field somewhere telling the auth service whether the client is a proxy tunnel. I'll have to look into this, thanks for pointing it out

File: rfd/0090-db-mfa-sessions.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (Joerger) [reply rfd/0090-db-mfa-sessions.md:25]: I see, I naively thought `tsh proxy db` worked the same as `tsh proxy ssh`, which prepares a single proxied connection directly to the node, rather than supporting multiple concurrent connections (I think).

I think it's still worth exploring maintaining the 1 minute TTL if possible, otherwise the security implications of per-session MFA are not consistent across Teleport protocols. If you've already exhausted potentials options and it isn't possible, then I don't intend to block this RFD or anything, I just wanted to make myself available for brainstorming here.

Another potential option could be to change how we enforce per-session MFA. You could perform the per-session MFA reissue and enforcement on `tsh proxy db --tunnel`, which will immediately setup the proxy connection with the 1-minute TTL cert. Then, you could let each db connection through the tunnel use non-mfa certs, and enforce per-session MFA on the db service side by checking that _either_ the connection has per-session MFA certs, or it was proxied from a proxy connection that does have per-session MFA certs.

File: rfd/0090-db-mfa-sessions.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (Joerger) [reply rfd/0090-db-mfa-sessions.md:74]: FWIW a solution like that could point towards my other comment - changing the way we enforce per-session MFA for proxy tunnel connections.

File: rfd/0090-db-mfa-sessions.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (GavinFrazar) [reply rfd/0090-db-mfa-sessions.md:25]: > I see, I naively thought tsh proxy db worked the same as tsh proxy ssh, which prepares a single proxied connection directly to the node, rather than supporting multiple concurrent connections (I think).

Yes, to clarify if it's not clear: `tsh proxy db --tunnel` starts a foreground process that accepts and proxies connections concurrently.

> you could let each db connection through the tunnel use non-mfa certs

Currently the `--tunnel` option allows local connections to connect without any certs at all, since they go through a TLS tunnel setup by the local proxy. The main use-case for this proxy tunnel is to enable GUI clients to connect and the local proxy handles ALPN/SNI TLS. The problem is that GUI clients create new connections all the time, and don't consistently re-use an existing connection. This was especially problematic with MS SQL server.

I think you're suggesting we make the local proxy establish a long-lived TLS connection and multiplex incoming client connections through that? That could work, except that users would need to also disable `disconnect_expired_cert` for DB resources, or the connection would die every minute. I think actually implementing this multiplexing would very difficult though, no? I think we would need a new ALPN wrapper like we did for the "-ping" variant

**Review by r0mant — APPROVED** [review]

File: rfd/0090-db-mfa-sessions.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (GavinFrazar) [reply rfd/0090-db-mfa-sessions.md:74]: @smallinsky We discussed ignoring this issue, but before I mark this resolved I just want to be sure that you agree.  If we want to address the issue, then we would need to add the version of `tsh` sent in `UserCertsRequest` (I think, unless there is some other way to determine the version of `tsh` the client is using).

[the remark this answers is not in the document]
Comment (GavinFrazar) [reply rfd/0090-db-mfa-sessions.md:74]: @Joerger revisiting this - instead of making`tsh db login/connect` return an error when `per_session_mfa` is enabled, we can just enforce the 1 minute TTL (and print a warning/suggestion to use `tsh proxy db --tunnel` instead for long lived sessions?). The proxy tunnel can have a new cert request parameter like `DisablePerSessionMFATTL = true` and then all proxy tunnels going forward from v12.x will retain long-lived database certs in memory, while all other use-cases will have the 1 minute TTL applied. What do you think?

[the remark this answers is not in the document]
Comment (GavinFrazar) [reply rfd/0090-db-mfa-sessions.md:74]: @smallinsky @r0mant @zmb3  your thoughts as well please. I will need to amend the RFD draft if this is what we'd like to do to address older tsh clients. I unfortunately did not have time to get this done in v12 and hadn't revisited this RFD until now.

With this approach we don't disable any existing login flows (as originally proposed) and still get the benefit that GUI clients will not cause MFA prompt spam.

[the remark this answers is not in the document]
Comment (smallinsky) [reply rfd/0090-db-mfa-sessions.md:74]: I think that we have discussed  this but I can't find any reference. 
The the question is why why should prevent the older clients  to use previous behaviour ? The user can still alter the tsh code and save the db cert on disc from memory also any mechanism based on user param can by bypassed.

[the remark this answers is not in the document]
Comment (Joerger) [reply rfd/0090-db-mfa-sessions.md:74]: My point here is that this new flow makes it possible to issue MFA-verified certificates with no TTL (or a longer TTL based on `max_session_ttl`). These certificates can be used with old/modified clients to bypass the expected MFA checks of DB access in general. They can also be potentially stolen by attackers to provide MFA-verified access past the 1-minute TTL, potentially with no TTL? 

The only way to prevent this downgrade in security is to have the Auth server enforce it. Right now, this is done by issuing MFA certificates with a 1-minute TTL. I'm not sure I have a good solution for replacing the 1-minute TTL...maybe we could add a grace period specifically for DB access through tunnel to allow recently expired MFA certificates, so that this doesn't affect other flows? Something server-enforced would be safer, so a client can't dictate the TTL of MFA certs themselves.

In my opinion, we should keep the default settings the most secure, so removing the 1-minute TTL by default seems a bit scary to me. My recommendation with the current design would be to add a new flag to configure this TTL, like `require_session_mfa_ttl` or `require_db_session_mfa_ttl`.

I just wanted to point out this change given that Per-session MFA is a security-focused features. If we think this isn't a significant security downgrade or is too theoretical, we can resolve this issue. Curious what others think.

[the remark this answers is not in the document]
Comment (GavinFrazar) [reply rfd/0090-db-mfa-sessions.md:74]: @Joerger totally understand the concern. TBH I have been nervous about actually implementing this feature for the same reasons (even though we've already approved the RFD).

If we let the client disable the 1-minute TTL, then only newer clients will get certs without the 1 minute TTL. Additionally, the newer clients will only disable that 1-minute TTL inside of the alpn local proxy tunnel, which holds the certs in memory only. So no version of tsh should save these longer-lived MFA certs to disk.

Now that being said, I have asked myself how customers would feel about this. Technically, a user in their cluster could build their own modified tsh that saves the certs to disk. Why they would want to do that - I don't know. But it would be possible. I can't see a malicious motive for doing that, as they already have the MFA token and login info to use their own account in the cluster. Perhaps they have some "innocent" reason for doing that - in which case the security posture of the cluster is weakened because users could "innocently" start saving certs to disk with their own modified tsh, and those certs could be exfiltrated by an attacker.

I think if that's a security downgrade that a customer finds unacceptable, then we could make this feature opt-in with configuration settings. For example:
```yaml
spec:
  require_session_mfa: true
  # defaults to 0h1m0s if not specified
  session_mfa_ttl: 0h5m0s # limit per-session MFA cert TTL to 5 minutes.
  # controls enforcement of per-session MFA cert TTL limit.
  # options are: "client", "on", "off"
  #   - client: allows clients to specify if requested MFA certs should be limited to `session_mfa_ttl`
  #   - on: enforce `session_mfa_ttl` max cert expiry always for per-session mfa certs.
  #   - off: disable enforcement of `session_mfa_ttl`
  session_mfa_ttl_enforcement: "client"
```

Benefit of adding configuration: allow customers to choose settings that fit their threat model.
Drawback: Even more configuration complexity (negative impact for time-to-value)

If we like the extra-configuration path, I'm open to suggestions around naming things 😅

[the remark this answers is not in the document]
Comment (GavinFrazar) [reply rfd/0090-db-mfa-sessions.md:74]: and of course we can get even more specific with the configuration, based on the type of session (db/app/ssh/etc). I'm only wary of adding further complexity to Teleport config.

[the remark this answers is not in the document]
Comment (GavinFrazar) [reply rfd/0090-db-mfa-sessions.md:74]: > They can also be potentially stolen by attackers to provide MFA-verified access past the 1-minute TTL, potentially with no TTL?

Regarding "potentially no TTL?": `max_session_ttl` is always enforced, and we have a maximum configurable limit on that, IIRC 30 hours. So Teleport will never issue any kind of cert without an expiration.

[the remark this answers is not in the document]
Comment (GavinFrazar) [reply rfd/0090-db-mfa-sessions.md:74]: I also think it's relevant to this discussion to add that we recently changed the interaction between `disconnect_expired_certs` and `require_session_mfa` TTL limits: https://github.com/gravitational/teleport/pull/18607

If `disconnect_expired_certs` is enabled, we now only disconnect certs based on the `tsh login` cert TTL rather than the 1-minute MFA session TTL. This (partially) helps the issue with DB GUI clients.

[the remark this answers is not in the document]
Comment (r0mant) [reply rfd/0090-db-mfa-sessions.md:74]: @GavinFrazar @Joerger @smallinsky This was discussed on the product sync quite a long time ago and we agreed upon treating a single Teleport Connect session (which effectively equates to "tsh proxy db") as MFA session to fix GUI clients UX. We'll need to extend this approach to Kube Access as well in Q1.
